### PR TITLE
osd/scrub: check reservation replies for relevance

### PR DESCRIPTION
**Pull request by ronen-fr**
Compare a token (nonce) carried in the reservation reply with the remembered
token of the reservation request.  If they don't match, the reply is stale and should
be ignored (and logged).

Fixes: https://tracker.ceph.com/issues/64052

### CODE REVIEW

**Comment by ronen-fr** [comment]
@athanatos - fixed now, and ready for review

File: src/messages/MOSDScrubReserve.h [hunk 1]
```diff
@@ -58,20 +62,21 @@ class MOSDScrubReserve : public MOSDFastDispatchOp {
...
+	break;
+      case GRANT:
+	out << "GRANT ";
+	break;
+      case REJECT:
+	out << "REJECT ";
+	break;
+      case RELEASE:
+	out << "RELEASE ";
+	break;
```
Comment (athanatos) [inline src/messages/MOSDScrubReserve.h:76]: Unrelated whitespace changes.

File: src/messages/MOSDScrubReserve.h [hunk 2]
```diff
@@ -58,20 +62,21 @@ class MOSDScrubReserve : public MOSDFastDispatchOp {
...
     }
-    out << "e" << map_epoch << ")";
+    out << "e" << map_epoch << " from: " << from
+	<< " req-id: " << reservation_nonce << ")";
```
Comment (athanatos) [inline src/messages/MOSDScrubReserve.h:79]: reservation_nonce:

File: src/osd/scrubber/scrub_reservations.cc [hunk 3]
```diff
@@ -169,23 +204,24 @@ bool ReplicaReservations::send_next_reservation_or_complete()
...
+		  m_sorted_secondaries.size(), ep, m_last_request_sent_nonce)
 	   << dendl;
   m_next_to_request++;
   return false;
 }
 
-void ReplicaReservations::verify_rejections_source(
-    OpRequestRef op,
-    pg_shard_t from)
+bool ReplicaReservations::handle_rejection(OpRequestRef op, pg_shard_t from)
```
Comment (athanatos) [inline src/osd/scrubber/scrub_reservations.cc:224]: This doesn't "handle" the request, it's a predicate checking recency and origin.  Rename it to is_rejection_valid and modify the signature to take a reference to the message type rather than an OpRequestRef.

File: src/osd/scrubber/scrub_reservations.cc [hunk 4]
```diff
@@ -231,4 +277,4 @@ std::ostream& ReplicaReservations::gen_prefix(
 	 << fmt::format("scrubber::ReplicaReservations:{}: ", fn);
 }
 
-} // namespace Scrub
+}  // namespace Scrub
```
Comment (athanatos) [inline src/osd/scrubber/scrub_reservations.cc:280]: Unrelated whitespace change.  Please take the time to check your PRs before submitting them.

File: src/osd/scrubber/scrub_reservations.cc [hunk 5]
```diff
@@ -194,20 +230,30 @@ void ReplicaReservations::verify_rejections_source(
...
 	   << dendl;
 
+  // is this a stale response to a previous request (e.g. one that
+  // timed-out)? if so - ignore it (but do log an error, to cause tests
+  // to fail)
+  const auto msg_epoch = op->get_req<MOSDScrubReserve>()->map_epoch;
+  const auto msg_nonce = op->get_req<MOSDScrubReserve>()->reservation_nonce;
+  const auto nonce_verified = is_response_relevant(msg_epoch, msg_nonce, from);
+  if (!nonce_verified) {
+    m_osds->clog->warn() << nonce_verified.error();
```
Comment (athanatos) [inline src/osd/scrubber/scrub_reservations.cc:240]: This shouldn't be a warning -- it's a normal if rare component of the protocol.  If the cancellation can't order past the original request, I think it'll actually happen every time.

File: src/osd/scrubber/scrub_reservations.cc [hunk 6]
```diff
@@ -194,20 +230,30 @@ void ReplicaReservations::verify_rejections_source(
...
-
-  ceph_assert(get_last_sent().has_value());
-  const auto expected = *get_last_sent();
-  if (from != expected) {
-    dout(1) << fmt::format(
-		   "unexpected rejection from {} (expected {})", from, expected)
-	    << dendl;
+  const auto peer_verified = is_msg_source_correct(msg_epoch, from);
+  if (!peer_verified) {
+    m_osds->clog->warn() << peer_verified.error();
```
Comment (athanatos) [inline src/osd/scrubber/scrub_reservations.cc:251]: This should actually be impossible, right?  If so, please explain that in the comment.

File: src/osd/scrubber/scrub_reservations.h [hunk 7]
```diff
@@ -44,6 +47,20 @@ namespace Scrub {
  *  that have been acquired until that moment.
  *  (Why? because we have encountered instances where a reservation request was
  *  lost - either due to a bug or due to a network issue.)
+ *
+ * Keeping primary & replica in sync:
+ *
+ *  Reservation requests are unique in that they may time-out and cancelled
+ *  by the primary. A request that was delayed for too long and was
```
Comment (athanatos) [inline src/osd/scrubber/scrub_reservations.h:54]: Close.

```
Reservation requests may be canceled by the primary independently of the replica's response.
Depending on timing, a cancellation by the primary might or might not be processed by a replica
prior to sending a response (either rejection or success).  Thus, we associate each reservation
request with a nonce incremented with each reservation during an interval and drop any responses
that do not match our current nonce -- this behavior is occurs after rejecting any messages from
prior intervals so reusing nonces between intervals is not a problem.  Note that epoch would not
suffice as it's possible for this sequence to occur several times without a new map epoch.
```

In writing the above, I have a few questions:
- is it actually possible for the cancellation to order past the original request?  Or is it the case that we're actually guaranteed to see the response from the replica one way or the other?
- if we see a success response from the replica, we can assume that it saw or will see the cancellation, right?

File: src/osd/scrubber/scrub_reservations.h [hunk 8]
```diff
@@ -64,6 +81,13 @@ class ReplicaReservations {
   /// for logs, and for detecting slow peers
   ScrubTimePoint m_last_request_sent_at;
 
+  /// a token to uniquely identify the last reservation request sent.
+  /// The response received will be checked against this nonce, to detect
+  /// stale responses.
+  /// Starts at 1, and is incremented for each request sent. '0' is reserved
+  /// for legacy messages.
```
Comment (athanatos) [inline src/osd/scrubber/scrub_reservations.h:88]: This comment doesn't really explain why the mechanism is necessary or sufficient.

```
/**
 * m_last_request_sent_nonce
 *
 * The primary is permitted to cancel outstanding reservation requests without
 * waiting for the pending response from the replica.  Thus, we may, in general,
 * see responses from prior reservation attempts that we need to ignore.  Each
 * reservation request is therefore associated with a nonce incremented within
 * an interval with each reservation request.  Any response with a non-matching
 * nonce must be from a reservation request we canceled.  Note that this check
 * occurs after validating that the message is from the current interval, so reusing
 * nonces between intervals is safe.
 *
 * 0 is a special value used to indicate that the sender did not include a nonce due
 * to not being a sufficiently recent version.
 */
```

File: src/osd/scrubber/scrub_reservations.h [hunk 9]
```diff
@@ -140,6 +168,17 @@ class ReplicaReservations {
...
   bool send_next_reservation_or_complete();
 
+  /// (internal helper) is this is a reply to our last request?
+  tl::expected<bool, std::string> is_response_relevant(
+      epoch_t msg_epoch,
+      reservation_nonce_t msg_nonce,
+      pg_shard_t from) const;
+
+  /// (internal helper) is this reply coming from the expected replica?
+  tl::expected<bool, std::string> is_msg_source_correct(
```
Comment (athanatos) [inline src/osd/scrubber/scrub_reservations.h:178]: The intended design should make this impossible, right?  That needs to be explained here.

File: src/osd/scrubber/scrub_reservations.h [hunk 10]
```diff
@@ -140,6 +168,17 @@ class ReplicaReservations {
    */
   bool send_next_reservation_or_complete();
 
+  /// (internal helper) is this is a reply to our last request?
```
Comment (athanatos) [inline src/osd/scrubber/scrub_reservations.h:171]:
```
/**
 * is_reservation_response_relevant
 *
 * Checks response once against m_last_request_sent_nonce -- see m_last_request_sent_nonce for details.
 */
```

File: src/osd/scrubber/scrub_reservations.h [hunk 11]
```diff
@@ -93,15 +117,19 @@ class ReplicaReservations {
...
   /**
+   * React to an incoming reservation rejection.
+   *
    * Verify that the sender of the received rejection is the replica we
-   * were expecting a reply from.
-   * If this is so - just mark the fact that the specific peer need not
-   * be released.
+   * were expecting a reply from, and that the message isn't stale.
+   * If a real rejection: log it, and mark the fact that the specific peer
+   * need not be released.
```
Comment (athanatos) [inline src/osd/scrubber/scrub_reservations.h:125]: "real rejection" and "stale" aren't useful unless you already know how this mechanism works.  Refer the reader to comments explaining the mechanisms.

File: src/osd/scrubber/scrub_reservations.h [hunk 12]
```diff
@@ -148,4 +187,4 @@ class ReplicaReservations {
   void log_success_and_duration();
 };
 
-} // namespace Scrub
+}  // namespace Scrub
```
Comment (athanatos) [inline src/osd/scrubber/scrub_reservations.h:190]: Whitespace.

File: src/osd/scrubber/scrub_reservations.cc [hunk 13]
```diff
@@ -125,16 +124,52 @@ ReplicaReservations::~ReplicaReservations()
...
+    pg_shard_t from) const
+{
+  const auto exp_source = get_last_sent();
+  if (!exp_source || from != *exp_source) {
+    return tl::unexpected(fmt::format(
+	"unexpected response from {} (with e:{}) (expected {})", from,
+	msg_epoch, exp_source.value_or(pg_shard_t{})));
+  }
+  return true;
+}
```
Comment (athanatos) [inline src/osd/scrubber/scrub_reservations.cc:152]: These helpers are more confusing than helpful.  Including msg_epoch in order to populate the error message suggests to the reader that it's relevant to the predicate.  I'd rather that they be pure predicates (returning bool) with the caller producing the error message.

File: src/osd/scrubber/scrub_machine.cc [hunk 14]
```diff
@@ -277,18 +277,25 @@ sc::result ReservingReplicas::react(const ReplicaReject& ev)
...
 
-  // manipulate the 'next to reserve' iterator to exclude
-  // the rejecting replica from the set of replicas requiring release
-  session.m_reservations->verify_rejections_source(ev.m_op, ev.m_from);
+  // Verify that the message is from the replica we were expecting a reply from,
+  // and that the message is not stale. If all is well - this is a real rejection:
+  // - log required details;
+  // - manipulate the 'next to reserve' iterator to exclude
+  //   the rejecting replica from the set of replicas requiring release
+  if (session.m_reservations->handle_rejection(ev.m_op, ev.m_from)) {
```
Comment (athanatos) [inline src/osd/scrubber/scrub_machine.cc:286]: Seems like it would probably be easier for this to just be

```
if (session.m_reservations->handle_rejection(ev.m_op, ev.m_from)) {
     return discard_event();
}
```

Though, handle_rejection should probably be something like is_reservation_response_valid.

**Review by athanatos — CHANGES_REQUESTED** [review]

File: src/osd/scrubber/scrub_reservations.cc (hunk 3 above)
Comment (athanatos) [reply src/osd/scrubber/scrub_reservations.cc:224]: Also, we have to apply the same check to a success, right?

Comment (ronen-fr) [reply src/osd/scrubber/scrub_reservations.cc:224]: > This doesn't "handle" the request, it's a predicate checking recency and origin

It's more than that now: previously, the FSM called both log_failure_and_duration() and verify_rejections_source(). Now - handle_rejection() does both (as there was no reason to require a separate API to ReservingReplica). If you do not like the generic 'handle' - maybe check_and_log_rejection()?

Comment (ronen-fr) [reply src/osd/scrubber/scrub_reservations.cc:224]: > signature to take a reference to the message type rather than an OpRequestRef.

done

Comment (ronen-fr) [reply src/osd/scrubber/scrub_reservations.cc:224]: > Also, we have to apply the same check to a success, right?

we do ?

File: src/osd/scrubber/scrub_reservations.cc (hunk 4 above)
Comment (ronen-fr) [reply src/osd/scrubber/scrub_reservations.cc:280]: I actually noticed those changes (introduced by the auto-formatting). I didn't think they were numerous
enough to warrant a separate commit. Will fix.

File: src/osd/scrubber/scrub_reservations.cc (hunk 5 above)
Comment (ronen-fr) [reply src/osd/scrubber/scrub_reservations.cc:240]: Good point.
I'll remove the warning. The idea was to notice such events, to detect possible bugs that would cause this scenario other than the known scenario of a very very delayed connection.

File: src/osd/scrubber/scrub_reservations.cc (hunk 6 above)
Comment (ronen-fr) [reply src/osd/scrubber/scrub_reservations.cc:251]: added a comment line

File: src/osd/scrubber/scrub_reservations.cc (hunk 3 above)
Comment (athanatos) [reply src/osd/scrubber/scrub_reservations.cc:224]: Oh, I see.  ScrubMachine gets distinct ReplicaReject and ReplicaGrant messages, so we can't really check the nonce in a single place -- that's unfortunate but would require more refactoring than is really worthwhile here to correct.

handle_rejection does have a state change I missed -- how about handle_reserve_rejection to make it symmetric to handle_reserve_grant?

File: src/osd/scrubber/scrub_reservations.h (hunk 7 above)
Comment (athanatos) [reply src/osd/scrubber/scrub_reservations.h:54]: If the replica goes down, there will be a new message once it is detected (including if it goes down and comes back up) resulting in an interval change.  As long as the primary is still in the same interval, it may safely assume that every message will be delivered to a replica (any exception would be a serious bug).

File: src/osd/scrubber/scrub_reservations.cc (hunk 6 above)
Comment (athanatos) [reply src/osd/scrubber/scrub_reservations.cc:251]: Where?  Having trouble finding it.

Comment (ronen-fr) [reply src/osd/scrubber/scrub_reservations.cc:251]: Sorry - I wasn't clear. I am making changes and acknowledging your comments one by one. Will push a new version today.

File: src/osd/scrubber/scrub_reservations.h (hunk 7 above)
Comment (ronen-fr) [reply src/osd/scrubber/scrub_reservations.h]: > so reusing nonces between intervals is not a problem

That comment made be aware of the fact that I was resetting the nonce too often. It is now moved to the
outer FSM state.

File: src/osd/scrubber/scrub_reservations.cc (hunk 6 above)
Comment (athanatos) [reply src/osd/scrubber/scrub_reservations.cc]: No worries, I'll check back in tomorrow.

**Comment by ronen-fr** [comment]
jenkins test make check

**Comment by ronen-fr** [comment]
jenkins test make check

File: src/osd/scrubber/scrub_reservations.cc (hunk 6 above)
Comment (ronen-fr) [reply src/osd/scrubber/scrub_reservations.cc]: @athanatos - I think I got all fixes right now.

File: src/osd/scrubber/scrub_machine.cc [hunk 15]
```diff
@@ -257,15 +259,17 @@ ReservingReplicas::~ReservingReplicas()
   // it's OK to try and clear the flag even if we don't hold it
   // (the flag remembers the actual holder)
   scrbr->clear_reserving_now();
+  // ignore any late-arriving reservation replies
+  context<PrimaryActive>().last_request_sent_nonce++;
```
Comment (athanatos) [inline src/osd/scrubber/scrub_machine.cc:263]: This already gets incremented in ReplicaReservations::send_next_reservation_or_complete().  Either strategy could actually work:
- doing it here would mean that a nonce is associated with a "round" of reservations and is incremented once complete or failed.
- doing it in ReplicaReservations::send_next_reservation_or_complete() would mean that a nonce is associated with a single request.

The latter gets my vote due to not needing to explain a "round", and the comment on last_request_sent_nonce indicates "reservation request is therefore associated with a nonce incremented within an interval with each reservation request".  Either way, let's not not do it in both places -- it'll be confusing.

**Review by athanatos — APPROVED** [review]
LGTM other than the comment about the extra nonce increment.

File: src/osd/scrubber/scrub_machine.cc (hunk 15 above)
Comment (ronen-fr) [reply src/osd/scrubber/scrub_machine.cc:263]: @athanatos,
I will remove it. It was added to serve as an additional guard before "harsh reactions" to
requests arriving from the wrong source:

both in handle_reserve_rejection() (scrub_reservations.cc:259) and in handle_reserve_grant()
(scrub_reservations.cc:169) we first check the nonce before checking (and possibly warning/aborting)
the source. The added increment was added to guarantee that for stale messages (from a previous
round altogether) we would not get a false positive.

As I do not see a sequence that would cause that, I will remove the extra increment.

**Comment by ronen-fr** [comment]
Merging based on multiple Teuthology batches
